js/src/data/api-fetch-middlewares.js

```diff
@@ -0,0 +1,56 @@
+/**
+ * Internal dependencies
+ */
+import { API_NAMESPACE } from './constants';
+
+/**
+ * @callback onErrorResponseCallback
+ * @param {Object} response The error response instance.
+ * @return {Promise<any>|any} Reject or throw the response error if want to continue original error handling process.
+ *                            Otherwise, resolve recovery result if the error could be processed here.
+ */
+
+/**
+ * Create a middleware with a callback to watch and handle the error response instance globally.
+ *
+ * @param  {onErrorResponseCallback} onErrorResponse A callback to be invoked when receiving error response.
+ * @return {import('@wordpress/api-fetch').APIFetchMiddleware} A middleware to watch and handle the error response instance globally.
+ */
+export function createErrorResponseCatcher( onErrorResponse ) {
+	const regexApiNamespace = new RegExp( `^${ API_NAMESPACE }\/` );
+
+	return function errorResponseCatcher( options, next ) {
+		// Requests issued from other places return and call the next middlewares with original options directly.
+		if ( ! regexApiNamespace.test( options.path ) ) {
+			return next( options );
+		}
+
+		const { parse: shouldParseResponse = true } = options;
+
+		return next( { ...options, parse: false } )
+			.catch( onErrorResponse )
+			.catch( ( response ) => {
+				if ( shouldParseResponse && response.json ) {
+					return response
+						.json()
+						.then( ( errMessage ) => Promise.reject( errMessage ) );
+				}
+
+				throw response;
+			} )
+			.then( ( response ) => {
```

Comment:
:nail_care: 
I think I found something that made me spend so much time parsing this in my head, trying to understand why this code is here and what it actually does. 

I hope the following comments could save that head-scratching for some other folks in the future.
```suggestion
		// Call onErrorResponse callback, then parse its fullfilment value to JSON if needed (and possible).
		return next( { ...options, parse: false } )
			.catch( onErrorResponse )
			.catch( ( response ) => {
				// Try parsing the rejection reason.
				// Reject with the response - parsed to JSON if required and possible.
				if ( shouldParseResponse && response.json ) {
					return response
						.json()
						.then( ( errMessage ) => Promise.reject( errMessage ) );
				}

				throw response;
			} )
			.then( ( response ) => {
				// Parse the "recovered" value.
```

Reply:
Also, on my way to the discovery of what those functions do I refactored the code a little. To something I personally find more readable [`fix/486-watch-401-error-parsing-helper`](https://github.com/woocommerce/google-listings-and-ads/compare/fix/486-watch-401-error...fix/486-watch-401-error-parsing-helper?expand=1) feel free to ignore it or merge it if you find any value in it.

Reply:
Thanks, I merged https://github.com/woocommerce/google-listings-and-ads/pull/834/commits/ec2593e55cd7a192721574889e3cea2d88442dad and appended two commits, https://github.com/woocommerce/google-listings-and-ads/pull/834/commits/0017500856474f5fc1e766034bcdc7899e214bb2 https://github.com/woocommerce/google-listings-and-ads/pull/834/commits/d114b76c180ac6bdf68a84c451018cad32e4b7ad, to tweak code comments.